DREAM/Optimization/tsgGradientDescent.hpp

```diff
@@ -0,0 +1,154 @@
+ * THE SOFTWARE WILL NOT INFRINGE ANY PATENT, COPYRIGHT, TRADEMARK, OR OTHER PROPRIETARY RIGHTS, OR THAT THE SOFTWARE WILL
+ * ACCOMPLISH THE INTENDED RESULTS OR THAT THE SOFTWARE OR ITS USE WILL NOT RESULT IN INJURY OR DAMAGE. THE USER ASSUMES
+ * RESPONSIBILITY FOR ALL LIABILITIES, PENALTIES, FINES, CLAIMS, CAUSES OF ACTION, AND COSTS AND EXPENSES, CAUSED BY, RESULTING
+ * FROM OR ARISING OUT OF, IN WHOLE OR IN PART THE USE, STORAGE OR DISPOSAL OF THE SOFTWARE.
+ */
+
+#ifndef __TASMANIAN_GRADIENT_DESCENT_HPP
+#define __TASMANIAN_GRADIENT_DESCENT_HPP
+
+#include "tsgOptimizationUtils.hpp"
+
+/*!
+ * \internal
+ * \file tsgGradientDescent.hpp
+ * \brief Gradient descent state and algorithm.
+ * \author Weiwei Kong & Miroslav Stoyanov
+ * \ingroup TasmanianOptimization
+ *
+ * Definition of the gradient descent state class and the particle swarm algorithm.
+ * \endinternal
+ */
+
+namespace TasOptimization {
+
+/*!
+ * \ingroup OptimizationState
+ * \brief Stores the information about a gradient descent run.
+ *
+ * \par class GradientDescentState
+ * An object of this class represents the components of a gradient descent run and all key parts of the swarm can be read through
+ * member methods. Methods are also provided to modify a subset of its parts.
+ *
+ * \par Constructors and Copy/Move assignment
+ * All constructors require information about an initial candidate point and initial stepsize. A method is available to retrieve
+ * the number of dimensions. No default constructor is available for this class, but copy/move constructors and assignment operator=
+ * overloads are available. The number of dimensions \b cannot be modified.
+ * - GradientDescentState()
+ * - getNumDimensions()
+ *
+ * \par Set and Get Candidate and Stepsize
+ * Throughout the run of gradient descent, the candidate point and stepsize may change. Methods are available to retrieve these
+ * data and set them manually.
+ * - getCandidate(), setCandidate()
+ * - getStepsize(), setStepsize()
+ *
+ * \par Gradient Descent Algorithm
+ * The gradient descent optimization algorithm can be used on a gradient descent algorithm with the goal of minimizing a particular
+ * objective function. The current candidate point, with respect to the given objective function, can be obtained from the state
+ * using the getCandidate() method.
+ * - GradientDescent()
+ *
+ * \par
+ * More information about the gradient descent algorithm can be found in the following paper:
+ *
+ * \par
+ * > Nesterov, Y. (2013). Gradient methods for minimizing composite functions. <em>Mathematical programming</em>, 140(1), 125-161.
+ *
+ */
+class GradientDescentState {
+  public:
+    //! \brief The default constructor is allowed.
+    GradientDescentState() = delete;
+    //! \brief Constructor for a gradient descent state with the initial candidate and stepsize.
+    GradientDescentState(const std::vector<double> &x, const double stepsize);
```

Comment:
I don't think the `stepsize` belongs in the state. I think this is a parameter that goes into the call to the descent method. It is common for descent methods to vary the `stepsize` or to adjust every few iterations. Even without an automated algorithm to adjust that, we should allow the user to make the adjustment.

Reply:
Also, why call it `x` when it should be `initial_iterate` or something related.